tensorboard/backend/event_processing/io_wrapper.py

```diff
@@ -163,6 +171,107 @@ def ListRecursivelyViaGlobbing(top):
         level += 1
 
 
+class ListRecursivelyViaBFSWalking:
+    """A class that walks a directory recursively with multiple threads.
+
+    For network based file systems, listing the directory might suffer from large latency.
+    This class walks the directory with breadth first strategy with concurrent threads to speed up.
+    The traverse result is put inside walked_structure.
+    Also, this class maintains a cache to speed up the query. If the cache is not expired,
+    then the class has no side effect.
+    Since I want to keep the function signature intact, some parameters are passed by global vars.
+    Args:
+      top_directory: A path to a directory.
+
+    Returns:
+      Nothing. The result is stored in return_structure. The return_structure is a list of
+      tuple(dir_path, dirs, filenames), where dirs (tuple) and filenames (tuple) are the directories
+      and files in dir_path.
+    """
+
+    class _BFSWALK:
```

Comment:
Personally I feel inner class is confusing; if we want global instance we can simply create a static/class function to instantiate the class